File: python/ultrasound_das/wave_animator.py
"""Wave propagation, delay-curve and before/after alignment computations.

Python port of ``wave_animator.m``. MATLAB is the source of truth.

Deliberate MATLAB<->Python divergences (keep both sides in sync with this in
mind, see ``CLAUDE.md``):

1. ``wave_animator.m`` mixes computation and MATLAB drawing in one file. Here
   the numeric part lives in this module (returning plain arrays) and the
   drawing lives in ``gui.py``. Port numeric changes here; port drawing
   changes to ``gui.py``.
2. ``wave_animator.m`` deliberately *duplicates* its own ``txArrivalTime``
   local function (see the comment at ``wave_animator.m:311``) to stay
   independent of the beamformer. This port instead reuses
   :func:`ultrasound_das.das_reference.tx_arrival_time`. The wavefront model
   is therefore guaranteed identical to the beamformer's by construction; if
   the MATLAB duplicate is ever changed without changing ``das_reference.m``,
   that divergence must be reproduced here on purpose.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from .das_reference import tx_arrival_time

__all__ = ["WavePropagation", "delay_curve_image", "alignment_bundles", "blue_white_red"]

_MAXSCAT = 30


def blue_white_red(n: int = 128) -> np.ndarray:
    """Blue-white-red diverging colormap as an (2n, 3) RGB array."""
    u = np.linspace(0.0, 1.0, n).reshape(-1, 1)
    lower = np.hstack([u, u, np.ones((n, 1))])
    upper = np.hstack([np.ones((n, 1)), np.flipud(u), np.flipud(u)])
    return np.vstack([lower, upper])


def _sample_rf(col, t0, fs, times):
    """Linear interpolation of one RF channel at arbitrary times (zero outside
    the record). Mirrors ``sampleRF``."""
    col = np.asarray(col, dtype=float)
    times = np.asarray(times, dtype=float).ravel()
    n = col.size
    pos = (times - t0) * fs                     # 0-based
    i0 = np.floor(pos).astype(np.int64)
    fr = pos - i0
    ok = (i0 >= 0) & (i0 < n - 1)
    i0c = np.clip(i0, 0, n - 2)
    v = col[i0c] * (1.0 - fr) + col[i0c + 1] * fr
    v[~ok] = 0.0
    return v


@dataclass
class WavePropagation:
    """Precomputed arrival-time maps for the wavefront animation, plus a
    :meth:`frame` method returning the field snapshot at time ``t``.

    Mirrors ``setupPropagation`` / ``drawPropagation`` (numeric parts only).
    """

    xv: np.ndarray
    zv: np.ndarray
    X: np.ndarray
    Z: np.ndarray
    TAU: np.ndarray
    c: float
    sigma: float
    xs: np.ndarray
    zs: np.ndarray
    rc: np.ndarray
    tau_s: np.ndarray
    Rmat: np.ndarray
    n_dropped: int
    elem_x: np.ndarray
    elem_z: np.ndarray
    act_elem_x: np.ndarray
    src_xz: tuple
    scheme: str
    tmax: float

    @classmethod
    def setup(cls, result, tx_idx, xrange, zrange, nx=241, nz=241):
        S = result
        tx = S.tx[tx_idx]
        c = S.c

        xv = np.linspace(xrange[0], xrange[1], nx)
        zv = np.linspace(max(zrange[0], 0.0), zrange[1], nz)
        X, Z = np.meshgrid(xv, zv)

        TAU = tx_arrival_time(X, Z, tx, c).reshape(X.shape, order="F")

        xs = np.asarray(S.scat["x"], dtype=float)
        zs = np.asarray(S.scat["z"], dtype=float)
        rc = np.asarray(S.scat["rc"], dtype=float)
        keep = rc != 0
        xs, zs, rc = xs[keep], zs[keep], rc[keep]
        n_dropped = 0
        if xs.size > _MAXSCAT:
            order = np.argsort(np.abs(rc))[::-1][:_MAXSCAT]
            n_dropped = xs.size - _MAXSCAT
            xs, zs, rc = xs[order], zs[order], rc[order]

        ns = xs.size
        tau_s = np.zeros(ns)
        Rmat = np.zeros((nz, nx, ns))
        for k in range(ns):
            tau_s[k] = float(tx_arrival_time(np.array([[xs[k]]]), np.array([[zs[k]]]), tx, c)[0])
            Rmat[:, :, k] = np.hypot(X - xs[k], Z - zs[k])

        fc = S.probe["fc"]
        bw = max(S.probe["bandwidth"], 1) / 100.0
        sigma = np.sqrt(2.0 * np.log(2.0)) / (np.pi * fc * bw)
        sigma = max(sigma, 0.3 / fc)

        elem_x = np.asarray(tx.elem_x, dtype=float)
        elem_z = np.asarray(tx.elem_z, dtype=float)
        act = np.isfinite(tx.delays) & (np.asarray(tx.apod) != 0)

        if ns > 0:
            dmax = np.max(np.hypot(xs[:, None] - elem_x[None, :], zs[:, None] - elem_z[None, :]), axis=1)
            tmax = float(np.max(tau_s + dmax / c) + 6.0 * sigma)
        else:
            tmax = float(np.max(TAU) + 6.0 * sigma)
        tmax = max(tmax, 6.0 * sigma)

        return cls(
            xv=xv, zv=zv, X=X, Z=Z, TAU=TAU, c=c, sigma=sigma,
            xs=xs, zs=zs, rc=rc, tau_s=tau_s, Rmat=Rmat, n_dropped=n_dropped,
            elem_x=elem_x, elem_z=elem_z, act_elem_x=elem_x[act],
            src_xz=tuple(tx.src_xz), scheme=tx.scheme, tmax=tmax,
        )

    def frame(self, t: float):
        """Field snapshot at time ``t`` [s]. Returns ``(field, hit_x)`` where
        ``field`` is in [-1, 1] (red = transmit wavefront, blue = echo) and
        ``hit_x`` are the x coordinates of receive elements the echo just
        reached."""
        sig = self.sigma
        F = np.exp(-0.5 * ((self.TAU - t) / sig) ** 2)

        Fe = np.zeros_like(F)
        hit_x = []
        for k in range(self.xs.size):
            dt = t - self.tau_s[k]
            if dt <= 0:
                continue
            amp = min(abs(self.rc[k]), 1.0) * 0.9
            Fe = np.maximum(Fe, amp * np.exp(-0.5 * ((self.Rmat[:, :, k] / self.c - dt) / sig) ** 2))
            re = np.hypot(self.elem_x - self.xs[k], self.elem_z - self.zs[k])
            hit_x.extend(self.elem_x[np.abs(re / self.c - dt) < 2.0 * sig].tolist())

        field = np.clip(F - Fe, -1.0, 1.0)
        return field, np.asarray(hit_x)


def delay_curve_image(result, tx_idx, tau):
    """Display data for the RF-image + delay-curve overlay.

    Returns ``(disp_rf, t_us, tau_us, ok)`` where ``disp_rf`` is the
    depth-gain-compensated RF image (for display only), ``t_us`` the time axis
    in microseconds, ``tau_us`` the per-element summed-sample time, and ``ok``
    the finite mask. Mirrors ``drawDelayCurve`` (data only)."""
    S = result
    RF = S.RF[:, :, tx_idx]
    t_us = (S.t0 + np.arange(RF.shape[0]) / S.fs) * 1e6
    denom = np.maximum(np.max(np.abs(RF), axis=1, keepdims=True), 1e-12) ** 0.7
    disp_rf = RF / denom
    tau = np.asarray(tau, dtype=float).ravel()
    ok = np.isfinite(tau)
    return disp_rf, t_us, tau * 1e6, ok


def alignment_bundles(result, tx_idx, tau, n_periods=5, n_samples=401, gain=3.0):
    """Before/after alignment waveform bundles.

    Returns a dict with ``trel_us``, ``pre`` [Nsamp x Nel], ``post``
    [Nsamp x Nel], ``ok`` mask, ``scale``, ``gain``, ``tc`` (window centre
    time), ``sum_trace`` (coherent sum of aligned active channels).
    Mirrors ``drawAlignment`` (data only)."""
    S = result
    RF = S.RF[:, :, tx_idx]
    nel = RF.shape[1]
    fc = S.probe["fc"]

    tau = np.asarray(tau, dtype=float).ravel()
    ok = np.isfinite(tau)
    if not np.any(ok):
        return None

    trel = np.linspace(-n_periods / fc, n_periods / fc, n_samples)
    tc = float(np.mean(tau[ok]))

    pre = np.zeros((trel.size, nel))
    post = np.zeros((trel.size, nel))
    for e in range(nel):
        pre[:, e] = _sample_rf(RF[:, e], S.t0, S.fs, tc + trel)
        if ok[e]:
            post[:, e] = _sample_rf(RF[:, e], S.t0, S.fs, tau[e] + trel)

    scale = max(np.max(np.abs(pre)), np.max(np.abs(post)), np.finfo(float).eps)
    sum_trace = np.sum(post[:, ok], axis=1)
    return dict(
        trel_us=trel * 1e6, pre=pre, post=post, ok=ok,
        scale=float(scale), gain=gain, tc=tc, sum_trace=sum_trace, tau=tau,
    )

File: wave_animator.m
function varargout = wave_animator(action, varargin)
%WAVE_ANIMATOR  波面伝搬・遅延曲線・整相前後比較の可視化ロジック。
%
%   MAIN_GUI から呼ばれる描画専用モジュール。状態は対象 axes の UserData に
%   保持するため、呼び出し側は axes ハンドルだけを持っていればよい。
%
%   ---- 波面アニメーション ---------------------------------------------
%   tmax = WAVE_ANIMATOR('setup_propagation', ax, S, txIdx, xlim_m, zlim_m)
%       送信波面と散乱エコーの場を計算する準備を行い、アニメーションの
%       最終時刻 [s] を返す。S は SIM_ENGINE の出力。
%   WAVE_ANIMATOR('draw_propagation', ax, t)
%       時刻 t [s] のスナップショットを描画する（CData の更新のみ）。
%
%   ---- 遅延曲線オーバーレイ -------------------------------------------
%   WAVE_ANIMATOR('delay_curve', ax, S, txIdx, tau, pt)
%       RF データを画像表示し、その上に「再構成点 pt に対して各素子の
%       どのサンプルが加算されるか」を遅延曲線として重ね描きする。
%       tau は [1 x Nel] の往復合計時間 [s]（DAS 実装の delays 出力の 1 行）。
%
%   ---- 整相前後の比較 -------------------------------------------------
%   WAVE_ANIMATOR('alignment', axPre, axPost, S, txIdx, tau, pt)
%       整相前のばらついた波形束（左）と、遅延補正で同相に揃った波形束＋
%       積算結果（右）を並べて描画する。
%
%   See also SIM_ENGINE, DAS_REFERENCE, MAIN_GUI.

switch lower(action)
    case 'setup_propagation'
        varargout{1} = setupPropagation(varargin{:});
    case 'draw_propagation'
        drawPropagation(varargin{:});
    case 'delay_curve'
        drawDelayCurve(varargin{:});
    case 'alignment'
        drawAlignment(varargin{:});
    otherwise
        error('wave_animator:action', '未知のアクション "%s" です。', action);
end
end

%% ======================================================================
function tmax = setupPropagation(ax, S, txIdx, xrange, zrange)
%SETUPPROPAGATION  波面／エコーの到達時刻マップを事前計算して描画枠を作る。
tx = S.tx(txIdx);
c  = S.c;

NX = 241; NZ = 241;
xv = linspace(xrange(1), xrange(2), NX);
zv = linspace(max(zrange(1), 0), zrange(2), NZ);
[X, Z] = meshgrid(xv, zv);

% 送信波面が各点に到達する時刻（DAS_REFERENCE と同一モデル）
TAU = txArrivalTime(X, Z, tx, c);
TAU = reshape(TAU, size(X));

% 散乱体ごとのエコー：到達時刻 + 散乱体からの距離 / c
%   スペックルファントムのように散乱体が多い場合は描画が破綻するので、
%   反射強度の大きい順に MAXSCAT 個だけをアニメーション対象にする。
MAXSCAT = 30;
xs = S.scat.x; zs = S.scat.z; rc = S.scat.rc;
keep = rc ~= 0;
xs = xs(keep); zs = zs(keep); rc = rc(keep);
nDropped = 0;
if numel(xs) > MAXSCAT
    [~, ord] = sort(abs(rc), 'descend');
    ord = ord(1:MAXSCAT);
    nDropped = numel(xs) - MAXSCAT;
    xs = xs(ord); zs = zs(ord); rc = rc(ord);
end
Ns = numel(xs);
tauS = zeros(1, Ns);
Rmat = zeros(NZ, NX, Ns);          % 散乱体からの距離を前計算（毎フレーム再利用）
for k = 1:Ns
    tauS(k) = txArrivalTime(xs(k), zs(k), tx, c);
    Rmat(:,:,k) = hypot(X - xs(k), Z - zs(k));
end

% パルス幅（可視化用の波面の厚み）
sigma = sqrt(2*log(2)) / (pi * S.probe.fc * max(S.probe.bandwidth,1)/100);
sigma = max(sigma, 0.3/S.probe.fc);

cla(ax, 'reset');
hold(ax, 'on');
st = struct();
st.X = X; st.Z = Z; st.TAU = TAU; st.c = c; st.sigma = sigma;
st.xs = xs; st.zs = zs; st.rc = rc; st.tauS = tauS;
st.Rmat = Rmat; st.nDropped = nDropped;
st.xv = xv; st.zv = zv;

% 場を表示する画像（赤 = 送信波面、青 = 散乱エコー）
st.hImg = imagesc(ax, xv*1e3, zv*1e3, zeros(NZ, NX));
colormap(ax, blueWhiteRed());
set(ax, 'CLim', [-1 1]);

% 素子・散乱体・焦点マーカー
plot(ax, tx.elem_x*1e3, tx.elem_z*1e3, 's', 'MarkerSize', 4, ...
     'MarkerEdgeColor', [.25 .25 .25], 'MarkerFaceColor', [.75 .75 .75], ...
     'HitTest','off');
act = isfinite(tx.delays) & tx.apod ~= 0;
plot(ax, tx.elem_x(act)*1e3, tx.elem_z(act)*1e3, 's', 'MarkerSize', 4, ...
     'MarkerEdgeColor', [.6 0 0], 'MarkerFaceColor', [1 .4 .4], 'HitTest','off');
st.hHit = plot(ax, nan, nan, 's', 'MarkerSize', 7, 'MarkerEdgeColor', [0 0 .6], ...
     'MarkerFaceColor', [.3 .6 1], 'HitTest','off');
if Ns > 0
    plot(ax, xs*1e3, zs*1e3, 'o', 'MarkerSize', 6, 'LineWidth', 1.2, ...
         'MarkerEdgeColor', [0 .5 0], 'MarkerFaceColor', 'none', 'HitTest','off');
end
if all(isfinite(tx.src_xz)) && ~strcmpi(tx.scheme,'plane')
    plot(ax, tx.src_xz(1)*1e3, tx.src_xz(2)*1e3, 'p', 'MarkerSize', 12, ...
         'MarkerEdgeColor', [.5 .3 0], 'MarkerFaceColor', [1 .8 .2], 'HitTest','off');
end

st.hTitle = title(ax, '');
axis(ax, 'image');
set(ax, 'YDir', 'reverse');
xlim(ax, xrange*1e3);
ylim(ax, [min(zv) max(zv)]*1e3);
xlabel(ax, 'x [mm]'); ylabel(ax, 'z [mm]');
grid(ax, 'off');
hold(ax, 'off');

% アニメーション終了時刻：最も遠い素子までエコーが戻りきる時刻
if Ns > 0
    dmax = max(hypot(xs(:) - tx.elem_x(:).', zs(:) - tx.elem_z(:).'), [], 2);
    tmax = max(tauS(:) + dmax/c) + 6*sigma;
else
    tmax = max(TAU(:)) + 6*sigma;
end
tmax = max(tmax, 6*sigma);
st.tmax = tmax;
st.elem_x = tx.elem_x; st.elem_z = tx.elem_z;
ax.UserData = st;
end

%% ======================================================================
function drawPropagation(ax, t)
%DRAWPROPAGATION  時刻 t のスナップショットに更新する。
st = ax.UserData;
if isempty(st) || ~isfield(st, 'hImg') || ~isvalid(st.hImg), return; end

sig = st.sigma;

% 送信波面（正 = 赤）
F = exp(-0.5*((st.TAU - t)/sig).^2);

% 散乱エコー（負 = 青）: 散乱体が照射された後に同心円状に広がる
%   重なった円弧が加算で飽和して送信波面を覆い隠さないよう、和ではなく
%   最大値で合成する。
Fe   = zeros(size(F));
hitx = [];
for k = 1:numel(st.xs)
    dt = t - st.tauS(k);
    if dt <= 0, continue; end
    amp = min(abs(st.rc(k)), 1) * 0.9;
    Fe  = max(Fe, amp * exp(-0.5*((st.Rmat(:,:,k)/st.c - dt)/sig).^2));

    % エコーが到達した受信素子をハイライト
    re = hypot(st.elem_x - st.xs(k), st.elem_z - st.zs(k));
    hitx = [hitx, st.elem_x(abs(re/st.c - dt) < 2*sig)]; %#ok<AGROW>
end

st.hImg.CData = max(min(F - Fe, 1), -1);
if isempty(hitx)
    set(st.hHit, 'XData', nan, 'YData', nan);
else
    set(st.hHit, 'XData', hitx*1e3, 'YData', zeros(size(hitx)));
end
if st.nDropped > 0
    extra = sprintf('  [表示は強い散乱体 %d 個のみ / 他 %d 個は非表示]', ...
                    numel(st.xs), st.nDropped);
else
    extra = '';
end
st.hTitle.String = sprintf('t = %.2f \\mus   （赤: 送信波面 / 青: 散乱エコー）%s', ...
                           t*1e6, extra);
end

%% ======================================================================
function drawDelayCurve(ax, S, txIdx, tau, pt)
%DRAWDELAYCURVE  RF 画像の上に、加算対象サンプルを示す遅延曲線を重ねる。
RF = S.RF(:,:,txIdx);
Nel = size(RF, 2);
tus = (S.t0 + (0:size(RF,1)-1)/S.fs) * 1e6;

cla(ax, 'reset');
% 深さ方向のゲイン補償をかけて浅部の飽和を防ぐ（表示のみ）
disp_rf = RF ./ max(max(abs(RF), [], 2), 1e-12).^0.7;
imagesc(ax, 1:Nel, tus, disp_rf);
colormap(ax, gray(256));
set(ax, 'CLim', [-1 1]*max(abs(disp_rf(:)))*0.6 + [-eps eps]);
hold(ax, 'on');

tauUs = tau(:).' * 1e6;
ok = isfinite(tauUs);
plot(ax, find(ok), tauUs(ok), '-', 'Color', [1 .25 .1], 'LineWidth', 2);
plot(ax, find(ok), tauUs(ok), '.', 'Color', [1 .8 0], 'MarkerSize', 8);
if any(~ok)
    yl = mean(tauUs(ok), 'omitnan');
    plot(ax, find(~ok), repmat(yl, 1, sum(~ok)), 'x', 'Color', [.4 .4 .4], ...
         'MarkerSize', 5);
end
hold(ax, 'off');
set(ax, 'YDir', 'normal');
xlabel(ax, '受信素子番号');
ylabel(ax, '時間 [\mus]');
if any(ok)
    span = max(max(tauUs(ok)) - min(tauUs(ok)), 0.5);
    ylim(ax, [min(tauUs(ok)) - 0.8*span - 1, max(tauUs(ok)) + 0.8*span + 1]);
end
xlim(ax, [0.5 Nel+0.5]);
title(ax, sprintf('遅延曲線  再構成点 (x, z) = (%.2f, %.2f) mm   有効素子 %d/%d', ...
      pt(1)*1e3, pt(2)*1e3, sum(ok), Nel));
end

%% ======================================================================
function drawAlignment(axPre, axPost, S, txIdx, tau, pt)
%DRAWALIGNMENT  整相前（左）と整相後＋積算（右）の波形束を描画する。
RF  = S.RF(:,:,txIdx);
Nel = size(RF, 2);
fc  = S.probe.fc;

tau = tau(:).';
ok  = isfinite(tau);
if ~any(ok)
    cla(axPre,'reset');  text(axPre, .5,.5,'有効な素子がありません','Units','normalized','HorizontalAlignment','center');
    cla(axPost,'reset'); text(axPost,.5,.5,'有効な素子がありません','Units','normalized','HorizontalAlignment','center');
    return;
end

trel = linspace(-5/fc, 5/fc, 401);          % 中心周波数の ±5 周期分
tc   = mean(tau(ok));

pre  = zeros(numel(trel), Nel);
post = zeros(numel(trel), Nel);
for e = 1:Nel
    pre(:,e) = sampleRF(RF(:,e), S.t0, S.fs, tc + trel);
    if ok(e)
        post(:,e) = sampleRF(RF(:,e), S.t0, S.fs, tau(e) + trel);
    end
end
scale = max([max(abs(pre(:))), max(abs(post(:))), eps]);
gain  = 3.0;

drawBundle(axPre,  trel*1e6, pre,  ok, scale, gain, tau, tc);
title(axPre, sprintf('整相前（共通時間窓 t = %.2f \\mus 中心）', tc*1e6));

drawBundle(axPost, trel*1e6, post, ok, scale, gain, [], []);
% 積算結果を波形束の下に重ねる
sumTrace = sum(post(:, ok), 2);
sScale = max(abs(sumTrace)) + eps;
hold(axPost, 'on');
yBase = -0.10*Nel - 2;
plot(axPost, trel*1e6, yBase + 0.9*(0.06*Nel+2)*sumTrace/sScale, ...
     '-', 'Color', [.85 .1 .1], 'LineWidth', 1.8);
plot(axPost, [trel(1) trel(end)]*1e6, [yBase yBase], ':', 'Color', [.6 .6 .6]);
text(axPost, trel(1)*1e6, yBase, sprintf('  \\Sigma  (peak %.3g)', max(abs(sumTrace))), ...
     'Color', [.85 .1 .1], 'VerticalAlignment', 'bottom', 'FontWeight', 'bold');
ylim(axPost, [yBase - (0.06*Nel+3), Nel + gain + 1]);
hold(axPost, 'off');
title(axPost, sprintf('整相後 + 積算   点 (%.2f, %.2f) mm', pt(1)*1e3, pt(2)*1e3));
end

%% ----------------------------------------------------------------------
function drawBundle(ax, tus, W, ok, scale, gain, tau, tc)
cla(ax, 'reset');
hold(ax, 'on');
Nel = size(W, 2);
for e = 1:Nel
    if ok(e)
        col = [0 .35 .75];
    else
        col = [.75 .75 .75];
    end
    plot(ax, tus, e + gain*W(:,e)/scale, '-', 'Color', col, 'LineWidth', 0.6);
end
if ~isempty(tau)
    % 整相前は各素子の「加算すべき時刻」がばらついていることを示す
    d = (tau - tc) * 1e6;
    plot(ax, d(ok), find(ok), '.', 'Color', [1 .3 0], 'MarkerSize', 9);
    plot(ax, d(ok), find(ok), '-', 'Color', [1 .3 0], 'LineWidth', 1.2);
else
    plot(ax, [0 0], [0 Nel+gain+1], '-', 'Color', [1 .3 0], 'LineWidth', 1.2);
end
hold(ax, 'off');
xlabel(ax, '相対時間 [\mus]');
ylabel(ax, '受信素子番号');
xlim(ax, [tus(1) tus(end)]);
ylim(ax, [0, Nel + gain + 1]);
grid(ax, 'on');
end

%% ----------------------------------------------------------------------
function v = sampleRF(col, t0, fs, times)
%SAMPLERF  RF の 1 チャンネルを任意時刻で線形補間する（範囲外は 0）。
n   = numel(col);
pos = (times(:) - t0)*fs + 1;
i0  = floor(pos);
fr  = pos - i0;
ok  = i0 >= 1 & i0 < n;
i0c = min(max(i0,1), n-1);
v   = col(i0c).*(1-fr) + col(i0c+1).*fr;
v(~ok) = 0;
end

%% ----------------------------------------------------------------------
function tau = txArrivalTime(X, Z, tx, c)
%TXARRIVALTIME  送信波面の到達時刻 [s]。
%  ※ DAS_REFERENCE の同名ローカル関数と同一モデル。可視化モジュールを
%    ビームフォーマから独立させるため意図的に複製している。
%    （「ビームフォーマが仮定している波面」をそのまま描くことが目的）
ex = tx.elem_x; ez = tx.elem_z; td = tx.delays; ap = tx.apod;
if isempty(ez), ez = zeros(size(ex)); end
act = isfinite(td) & ap(:).' ~= 0;
ex = ex(act); ez = ez(act); td = td(act);
xg = X(:); zg = Z(:);

src = tx.src_xz;
isFocused = strcmpi(tx.scheme,'focused') && numel(src)==2 && all(isfinite(src)) && ...
            src(2) > 0 && ~tx.single_element;
if isFocused
    Rf   = hypot(ex - src(1), ez - src(2));
    Tf   = mean(td + Rf/c);
    D    = hypot(src(1), src(2));
    u    = [src(1) src(2)]/D;
    proj = xg*u(1) + zg*u(2);
    dF   = hypot(xg - src(1), zg - src(2));
    sgn  = ones(numel(xg),1); sgn(proj <= D) = -1;
    tau  = Tf + sgn.*dF/c;
else
    tau = inf(numel(xg), 1);
    for e = 1:numel(ex)
        tau = min(tau, td(e) + hypot(xg - ex(e), zg - ez(e))/c);
    end
end
end

%% ----------------------------------------------------------------------
function cmap = blueWhiteRed()
%BLUEWHITERED  青-白-赤のダイバージングカラーマップ（ツールボックス不要）。
n = 128;
u = linspace(0, 1, n).';
lower_ = [u, u, ones(n,1)];                 % 青 -> 白
upper_ = [ones(n,1), flipud(u), flipud(u)]; % 白 -> 赤
cmap = [lower_; upper_];
end
